File: src/main/java/ast/ASTExprOp.java
package ast;

public class ASTExprOp extends ASTExpr {

    public enum ExprType {
        POSTFIX_INC,
        POSTFIX_DEC,
        PREFIX_INC,
        PREFIX_DEC,
        NEGATIVE,
        LOGICAL_NOT,
        BINARY_NOT,
        MULT,
        DIV,
        MOD,
        ADD,
        SUB,
        LSHFT,
        RSHFT,
        LT,
        LTE,
        GT,
        GTE,
        EQ,
        NEQ,
        BINARY_AND,
        BINARY_OR,
        BINARY_XOR,
        LOGICAL_AND,
        LOGICAL_OR

    }

    /* if the expr has one or two operands */
    public boolean is_unary;

    public ASTExpr left;
    public ASTExpr right;
    public ExprType type;

    public ASTExprOp(ASTFileLocation loc, ExprType type, ASTExpr left, ASTExpr right){
        super(loc);
        this.left = left;
        this.right = right;
        is_unary = false;
        this.type = type;

        if(type == ExprType.PREFIX_DEC || type == ExprType.PREFIX_INC
                || type == ExprType.POSTFIX_DEC || type == ExprType.POSTFIX_INC
                || type == ExprType.NEGATIVE || type == ExprType.LOGICAL_NOT
                || type == ExprType.BINARY_NOT){
            is_unary = true;
        }
    }

    public void print(int i){
        printIndent(i);
        System.out.print("ASTExprOp: ");
        System.out.println(type);
        if(left != null){
            left.print(i+1);
        }
        if(right != null){
            right.print(i+1);
        }
    }
}

File: src/main/kotlin/ast/ASTExprOp.kt
package ast

class ASTExprOp(loc: ASTFileLocation, val type: ExprType, val left: ASTExpr, val right: ASTExpr?) : ASTExpr(loc) {

    /* if the expr has one or two operands */
    var is_unary: Boolean = false

    enum class ExprType {
        POSTFIX_INC,
        POSTFIX_DEC,
        PREFIX_INC,
        PREFIX_DEC,
        NEGATIVE,
        LOGICAL_NOT,
        BINARY_NOT,
        MULT,
        DIV,
        MOD,
        ADD,
        SUB,
        LSHFT,
        RSHFT,
        LT,
        LTE,
        GT,
        GTE,
        EQ,
        BINARY_AND,
        BINARY_OR,
        BINARY_XOR,
        LOGICAL_AND,
        LOGICAL_OR,
        ARRAY;

        companion object {
            fun isUnary(type: ExprType): Boolean {
                return(type == PREFIX_DEC || type == PREFIX_INC
                        || type == POSTFIX_DEC || type == POSTFIX_INC
                        || type == NEGATIVE || type == LOGICAL_NOT
                        || type == BINARY_NOT)
            }

            /* logical ops require args converted to boolean first */
            fun isLogical(type: ExprType): Boolean {
                return type == LOGICAL_NOT || type == LOGICAL_AND || type == LOGICAL_OR
            }

            fun isNotOverridable(type: ExprType): Boolean {
                return isLogical(type) || type == EQ
            }
        }
    }

    init {
        is_unary = ExprType.isUnary(type)
    }

    override fun print(i: Int) {
        printIndent(i)
        print("ASTExprOp: ")
        println(type)
        left.print(i + 1)
        right?.print(i + 1)
    }

    override fun isLValue(): Boolean {
        return type == ExprType.ARRAY
    }
}